virtcontainers/qemu_ppc64le.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2018 IBM
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package virtcontainers
+
+import (
+	"encoding/hex"
+	"fmt"
+	"os"
+	"strconv"
+
+	govmmQemu "github.com/intel/govmm/qemu"
+)
+
+type qemuPPC64le struct {
+	// inherit from qemuArchBase, overwrite methods if needed
+	qemuArchBase
+}
+
+const defaultQemuPath = "/usr/bin/qemu-system-ppc64le"
+
+const defaultQemuMachineType = QemuPseries
+
+const defaultQemuMachineOptions = "accel=kvm,usb=off"
+
+const defaultPCBridgeBus = "pci.0"
+
+var qemuPaths = map[string]string{
+	QemuPseries: defaultQemuPath,
+}
+
+var kernelParams = []Param{
+	{"tsc", "reliable"},
+	{"no_timer_check", ""},
+	{"rcupdate.rcu_expedited", "1"},
+	{"i8042.direct", "1"},
+	{"i8042.dumbkbd", "1"},
+	{"i8042.nopnp", "1"},
+	{"i8042.noaux", "1"},
+	{"noreplace-smp", ""},
+	{"reboot", "k"},
+	{"console", "hvc0"},
+	{"console", "hvc1"},
+	{"iommu", "off"},
+	{"cryptomgr.notests", ""},
+	{"net.ifnames", "0"},
+	{"pci", "lastbus=0"},
+}
+
+var supportedQemuMachines = []govmmQemu.Machine{
+	{
+		Type:    QemuPseries,
+		Options: defaultQemuMachineOptions,
+	},
+}
+
+// returns the maximum number of vCPUs supported
+func maxQemuVCPUs() uint32 {
+	return uint32(128)
+}
+
+func newQemuArch(config HypervisorConfig) qemuArch {
+	machineType := config.HypervisorMachineType
+	if machineType == "" {
+		machineType = defaultQemuMachineType
+	}
+
+	return &qemuPPC64le{
+		qemuArchBase{
+			machineType:           machineType,
+			qemuPaths:             qemuPaths,
+			supportedQemuMachines: supportedQemuMachines,
+			kernelParamsNonDebug:  kernelParamsNonDebug,
+			kernelParamsDebug:     kernelParamsDebug,
+			kernelParams:          kernelParams,
+		},
+	}
+}
+
+func (q *qemuPPC64le) capabilities() capabilities {
+	var caps capabilities
+
+	// pseries machine type supports hotplugging drives
+	if q.machineType == QemuPseries {
+		caps.setBlockDeviceHotplugSupport()
+	}
+
+	return caps
+}
+
+func (q *qemuPPC64le) bridges(number uint32) []Bridge {
```

Comment:
There is quite a bit of commonality in the following functions between amd64+ppc64el:

- `bridges()`
- `memoryTopology()`
- `appendBridges()`

I'd say it's "borderline refactoring territory" but wdyt @sboeuf?

Reply:
@nitkon yes you could factorize more for those functions here. Correct me if I'm wrong but the code for ppc64 is a copy of the implementation for amd64 (for the 3 functions @jodh-intel mentioned). `bridges()` is slightly different because of the type itself, but the semantic is the same.

Reply:
This looks like the only outstanding issue stopping this from landing.

Reply:
@jodh-intel : Thanks for pointing it out. Will update my PR accordingly.

Reply:
@jodh-intel : @sboeuf : 
The code for ``` memoryTopology() ``` is not exactly the same for ppc64 and x86. The addition being in ppc64le
```//  Align hostMemoryMb to 256 MB multiples
hostMemoryMb -= (hostMemoryMb % 256)
```  

The code for ```appendBridges()```  has already been refactored. 
```
// appendBridges appends to devices the given bridges
func (q *qemuPPC64le) appendBridges(devices []govmmQemu.Device, bridges []Bridge) []govmmQemu.Device {
        return genericAppendBridges(devices, bridges, q.machineType)
}
  ```


``` bridges() ``` can be refactored furthure.

Reply:
Hi @nitkon - yes. I think all we really need now is something like the following (untested! :) that the arch-specific qemu functions can call:

```
func genericMakeBridges(number uint32, bt bridgeType) []Bridge {
        var bridges []Bridge

	for i := uint32(0); i < number; i++ {
		bridges = append(bridges, Bridge{
			Type:    bt,
			ID:      fmt.Sprintf("%s-bridge-%d", bt, i),
			Address: make(map[uint32]string),
		})
	}

        return bridges
}


func genericMemoryTopology(memoryMb, hostMemoryMb uint64, alignMB uint32) govmmQemu.Memory {
	// NVDIMM device needs memory space 1024MB
	memoryOffset := 1024

        if alignMB > 0 {
                hostMemoryMb -= (hostMemoryMb % alignMB)
        }

	// add 1G memory space for nvdimm device (vm guest image)
	memMax := fmt.Sprintf("%dM", hostMemoryMb+uint64(memoryOffset))

	mem := fmt.Sprintf("%dM", memoryMb)

	memory := govmmQemu.Memory{
		Size:   mem,
		Slots:  defaultMemSlots,
		MaxMem: memMax,
	}

	return memory
}
```